src/java.base/share/classes/jdk/internal/foreign/BufferStack.java

```diff
@@ -0,0 +1,188 @@
+ * questions.
+ */
+
+package jdk.internal.foreign;
+
+import jdk.internal.misc.CarrierThreadLocal;
+import jdk.internal.vm.annotation.ForceInline;
+
+import java.lang.foreign.Arena;
+import java.lang.foreign.MemoryLayout;
+import java.lang.foreign.MemorySegment;
+import java.lang.foreign.SegmentAllocator;
+import java.lang.ref.Reference;
+import java.util.Objects;
+import java.util.concurrent.locks.ReentrantLock;
+
+import static java.lang.foreign.ValueLayout.JAVA_LONG;
+
+/**
+ * A buffer stack that allows efficient reuse of memory segments. This is useful in cases
+ * where temporary memory is needed.
+ * <p>
+ * Use the factory {@link #of(long)} to create new instances of this class.
+ * <p>
+ * Note: The reused segments are neither zeroed out before nor after re-use.
+ */
+public record BufferStack(long size, CarrierThreadLocal<PerThread> tl) {
+
+    /**
+     * {@return a new Arena that tries to provide {@code size} and {@code byteAlignment}
+     *          allocations by recycling the BufferStacks internal memory}
+     *
+     * @param size          to be reserved from this BufferStacks internal memory
+     * @param byteAlignment to be used for reservation
+     */
+    @ForceInline
+    public Arena pushFrame(long size, long byteAlignment) {
+        return tl.get().pushFrame(size, byteAlignment);
+    }
+
+    /**
+     * {@return a new Arena that tries to provide {@code layout}
+     *          allocations by recycling the BufferStacks internal memory}
+     *
+     * @param layout for which to reserve internal memory
+     */
+    @ForceInline
+    public Arena pushFrame(MemoryLayout layout) {
+        return pushFrame(layout.byteSize(), layout.byteAlignment());
+    }
+
+    @Override
+    public String toString() {
+        return "BufferStack[" + size + "]";
+    }
+
+    @Override
+    public int hashCode() {
+        return System.identityHashCode(this);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        return this == obj;
+    }
+
+    private record PerThread(ReentrantLock lock, SlicingAllocator stack, Arena arena) {
+
+        @ForceInline
+        public Arena pushFrame(long size, long byteAlignment) {
+            boolean needsLock = Thread.currentThread().isVirtual() && !lock.isHeldByCurrentThread();
+            if (needsLock && !lock.tryLock()) {
+                // Rare: another virtual thread on the same carrier competed for acquisition.
+                return Arena.ofConfined();
+            }
+            if (!stack.canAllocate(size, byteAlignment)) {
+                if (needsLock) lock.unlock();
+                return Arena.ofConfined();
+            }
+            return new Frame(needsLock, size, byteAlignment);
+        }
+
+        static PerThread of(long size) {
+            final Arena arena = Arena.ofAuto();
+            final SlicingAllocator stack = new SlicingAllocator(arena.allocate(size));
+            return new PerThread(new ReentrantLock(), stack, arena);
```

Comment:
```suggestion
            final Arena arena = Arena.ofAuto();
            return new PerThread(new ReentrantLock(), 
                                 new SlicingAllocator(arena.allocate(size)), arena);
        }
```
The local variable stack is only used once, is it okay not to use it?